Iris/IrisCompiler/BackEnd/PeEmitter.cs

```diff
@@ -63,30 +64,66 @@ public void Flush()
                 _textEmitter.Dispose();
                 _textEmitter = null;
 
-                string platform = _flags.HasFlag(CompilationFlags.Platform32) ? " /32BITPREFERRED" : " /X64";
-                string debug = _flags.HasFlag(CompilationFlags.NoDebug) ? string.Empty : " /DEBUG";
-                string dll = _flags.HasFlag(CompilationFlags.WriteDll) ? " /DLL" : string.Empty;
+                string platform = _flags.HasFlag(CompilationFlags.Platform32) ? " -32BITPREFERRED" : " -X64";
+                string debug = _flags.HasFlag(CompilationFlags.NoDebug) ? string.Empty : " -DEBUG";
+                string dll = _flags.HasFlag(CompilationFlags.WriteDll) ? " -DLL" : string.Empty;
 
-                // Find the path to ilasm.exe
-                string mscorlibPath = typeof(object).Assembly.Location;
-                string frameworkDir = Path.GetDirectoryName(mscorlibPath);
-                string ilasmPath = Path.Combine(frameworkDir, "ilasm.exe");
+                string mscorlibPath;
+                string frameworkDir;
+                string ilasmPath = null;
+
+                bool isWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
+
+                // If we are on Windows, and we aren't targetting .NET Core, look next to our CoreLib's ilasm.exe. If we are running on desktop framework this will work.
+                if (!_flags.HasFlag(CompilationFlags.NetCore) && isWindows)
+                {
+                    mscorlibPath = typeof(object).Assembly.Location;
+                    frameworkDir = Path.GetDirectoryName(mscorlibPath);
+                    ilasmPath = Path.Combine(frameworkDir, "ilasm.exe");
+                }
+
+                if (ilasmPath == null || !File.Exists(ilasmPath))
+                {
+                    string thisDir = Path.GetDirectoryName(GetType().Assembly.Location);
+                    string ilasmFileName = isWindows ? "ilasm.exe" : "ilasm";
+                    ilasmPath = Path.Combine(thisDir, ilasmFileName);
+
+                    if (!File.Exists(ilasmPath))
+                        throw new FileNotFoundException("ilasm cannot be found, make sure netcore ilasm is present in the same directory as IrisCompiler.dll");
+
+                    if (!string.IsNullOrEmpty(debug)) // netcore ilasm requires explicitly specifying pdb format
```

Comment:
>) [](start = 52, length = 1)

Should this specify full PDBs if building for Desktop CLR?  AFAIK Roslyn creates full PDBs unless building for .NET Core.

Reply:
Prior to .NET 5.0 ilasm for .NET Core did not have support for producing PDB at all. 
With the 5.0 version it received the support for Portable PDB generation.
I've introduced the `-pdbfmt` flag to emphasize this but also to leave it open in case Microsoft decides to support other variants/formats (embedded, full, legacy-classic PDBs)

Reply:
FYI Patrick your statement isn't quite correct: Roslyn (or maybe it is actually msbuild targets?) *defaults* to full PDBs on desktop, but the options are independent. So it is possible to get a portable PDB when targeting desktop or a full PDB when targeting .NET Core (assuming you are on Windows).

Regardless, it would be very weird to get into this code path if you are targeting desktop since we will generally find ilasm.exe in the above code path in this case. So I don't think it makes sense to change it.

Reply:
Right, I see the code path shouldn't be used on desktop.

---
In reply to: [562864868](https://github.com/microsoft/ConcordExtensibilitySamples/pull/65#discussion_r562864868) [](ancestors = 562864868)